### PR TITLE
meson: Fix my mistakes

### LINKED ISSUE #86229: Meson: Unknown compiler for cross compilation
**Describe the bug**

At least the following derivations fail on current `nixos-unstable`:

```
nixpkgs.pkgsCross.aarch64-multiplatform.fuse
nixpkgs.pkgsCross.aarch64-multiplatform.xorg.xorgproto
nixpkgs.pkgsCross.aarch64-multiplatform.glib
```

All failures have this log in common:

```
meson.build:1:0: ERROR: Unknown compiler(s): ['cc', 'gcc', 'clang', 'pgcc', 'icc']
The follow exceptions were encountered:
Running "cc --version" gave "[Errno 2] No such file or directory: 'cc': 'cc'"
Running "gcc --version" gave "[Errno 2] No such file or directory: 'gcc': 'gcc'"
Running "clang --version" gave "[Errno 2] No such file or directory: 'clang': 'clang'"
Running "pgcc --version" gave "[Errno 2] No such file or directory: 'pgcc': 'pgcc'"
Running "icc --version" gave "[Errno 2] No such file or directory: 'icc': 'icc'"
```

**To Reproduce**

Build one of `fuse`, `xorg.xorgproto`, `glib` for `aarch64-multiplatform` on `nixos-unstable`.

**Expected behavior**
They used to build.

**Additional context**

Git bisect tells me, that one of these commits are to blame:

9073a0cb8b2f419785a60969e11e96733f29b200..6c162416ca4d95b54a9837293e7e0f1b01f4622c

I couldn't bisect further as that would have taken lots of build time.

**Notify maintainers**

As `meson` seems to be the common denominator, meson maintainers: @jtojnar @rasendubi @brandonedens 

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
  - nixpkgs.pkgsCross.aarch64-multiplatform.fuse
  - nixpkgs.pkgsCross.aarch64-multiplatform.xorg.xorgproto
  - nixpkgs.pkgsCross.aarch64-multiplatform.glib
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

Fixes #86229 and something that @matthewbauer noticed.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/build-managers/meson/default.nix [hunk 1]
```diff
@@ -58,6 +59,10 @@ python3Packages.buildPythonApplication rec {
 
   setupHook = ./setup-hook.sh;
 
+  # Ensure there will always be a native C compiler when meson is used, as a
+  # workaround until https://github.com/mesonbuild/meson/pull/6512 lands.
+  depsHostHostPropagated = [ pkgsHostHost.stdenv.cc ];
```
Comment (Ericson2314) [inline pkgs/development/tools/build-managers/meson/default.nix:64]: Something tells me this is the first `depsHostHostPropagated` :joy:

**Comment by samueldr** [comment]
I have bisected an issue with systemd's cross-compilation to this PR.

Since https://github.com/NixOS/nixpkgs/commit/05d26adb0ad90ec17ad38a0ed638e9be33cbc7c7, `nix-build -A pkgsCross.aarch64-multiplatform.systemd` fails with the following:

```
...
Option prefix is: /nix/store/0rjxdagi2y3crgf8a5n809zhwy5gwxrz-systemd-245.3-aarch64-unknown-linux-gnu [default: /usr]
Found ninja-1.10.0 at /nix/store/2cjrhrq96f4bb20yywnd4zgbqv87g21m-ninja-1.10.0/bin/ninja
Using 'STRIP' from environment with value: 'aarch64-unknown-linux-gnu-strip'
Using 'STRIP' from environment with value: 'aarch64-unknown-linux-gnu-strip'
building
build flags: -j12 -l12
[602/1257] Generating stub.so with a custom commandstemd_userdb-glue.c.o'.o'K[Kn-gperf.c.o'[KKo'o'
FAILED: src/boot/efi/stub.so
/nix/store/41vfz2mi98j2wkgqx6ggj71bl314f604-gcc-wrapper-9.3.0/bin/ld -o src/boot/efi/stub.so -T /nix/store/zp7a35566pc4vvq3rfnzm8f3xzadd2vk-gnu-efi-3.0.11-aarch64-unknown-linux-gnu/lib/elf_aarch64_efi.lds -shared -Bsymbolic -nostdlib -znocombreloc -L /nix/store/zp7a35566pc4vvq3rfnzm8f3xzadd2vk-gnu-efi-3.0.11-aarch64-unknown-linux-gnu/lib /nix/store/zp7a35566pc4vvq3rfnzm8f3xzadd2vk-gnu-efi-3.0.11-aarch64-unknown-linux-gnu/lib/crt0-efi-aarch64.o --defsym=EFI_SUBSYSTEM=0xa src/boot/efi/disk.c.o src/boot/efi/graphics.c.o src/boot/efi/measure.c.o src/boot/efi/pe.c.o src/boot/efi/util.c.o src/boot/efi/linux.c.o src/boot/efi/splash.c.o src/boot/efi/stub.c.o -lefi -lgnuefi /nix/store/jj72ah48qnr2qx8bawd62wj3s9fcfndm-aarch64-unknown-linux-gnu-stage-final-gcc-debug-9.3.0/lib/gcc/aarch64-unknown-linux-gnu/9.3.0/libgcc.a
/nix/store/bfi8cjnhmqs8nd79ryr9n23myvlwhc4l-binutils-2.34/bin/ld: cannot represent machine `aarch64'
[613/1257] Generating boot.c.o with a custom commandwn-core@sta/nspawn-oci.c.o'.o'
ninja: build stopped: subcommand failed.
builder for '/nix/store/fwzs1q8qvfjz43lmiaxs3addxls6rxx3-systemd-245.3-aarch64-unknown-linux-gnu.drv' failed with exit code 1
```

https://gist.github.com/da2b61d3f18656acf5a94a7f3193ceeb

With the important bit being <code>cannot represent machine `aarch64'</code>

* * *

With that said, this may not be the PR that introduced the actual regression, but only the PR that made it fail in this particular exact way. The cause may be deeper, and introduced whenever the issue that caused #86229 was introduced (introduced in 9073a0cb8b2f419785a60969e11e96733f29b200). Bisecting this does not look trivial.

Any idea?

**Comment by Ericson2314** [comment]
I've since fixed the issue. I wouldn't be surprised if it was still user the wrong `ld`, but somehow successfully, before---I can't imagine how before this fix the build was passing "for the right reasons".

**Pull request merged**